contracts/src/token/erc1155/extensions/supply.rs

```diff
@@ -0,0 +1,881 @@
+//! Extension of ERC-1155 that adds tracking of total supply per token id.
+//!
+//! Useful for scenarios where Fungible and Non-fungible tokens have to be
+//! clearly identified. Note: While a total_supply of 1 might mean the
+//! corresponding is an NFT, there are no guarantees that no other tokens
+//! with the same id are not going to be minted.
+//!
+//! NOTE: This contract implies a global limit of 2**256 - 1 to the number
+//! of tokens that can be minted.
+//!
+//! CAUTION: This extension should not be added in an upgrade to an already
+//! deployed contract.
+
+use alloc::{vec, vec::Vec};
+
+use alloy_primitives::{Address, U256};
+use stylus_sdk::{
+    abi::Bytes,
+    msg,
+    prelude::{public, sol_storage},
+    storage::TopLevelStorage,
+};
+
+use crate::{
+    token::erc1155::{self, Erc1155, IErc1155},
+    utils::math::storage::SubAssignUnchecked,
+};
+
+sol_storage! {
+    /// State of an [`Erc1155Supply`] token.
+    pub struct Erc1155Supply {
+        /// ERC-1155 contract storage.
+        Erc1155 erc1155;
+        /// Mapping from token id to total supply.
+        mapping(uint256 => uint256) _total_supply;
+        /// Total supply of all token ids.
+        uint256 _total_supply_all;
+    }
+}
+
+/// NOTE: Implementation of [`TopLevelStorage`] to be able use `&mut self` when
+/// calling other contracts and not `&mut (impl TopLevelStorage +
+/// BorrowMut<Self>)`. Should be fixed in the future by the Stylus team.
+unsafe impl TopLevelStorage for Erc1155Supply {}
+
+#[public]
+impl Erc1155Supply {
+    /// Total value of tokens in with a given id.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - Read access to the contract's state.
+    /// * `id` - Token id as a number.
+    pub fn total_supply(&self, id: U256) -> U256 {
+        self._total_supply.get(id)
+    }
+
+    /// Total value of tokens.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - Read access to the contract's state.
+    #[selector(name = "totalSupply")]
+    pub fn total_supply_all(&self) -> U256 {
+        *self._total_supply_all
+    }
+
+    /// Indicates whether any token exist with a given id, or not.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - Read access to the contract's state.
+    /// * `id` - Token id as a number.
+    pub fn exists(&self, id: U256) -> bool {
+        self.total_supply(id) > U256::ZERO
+    }
+
+    /// Returns the value of tokens of type `id` owned by `account`.
+    ///
+    /// Re-export of [`Erc1155::balance_of`]
```

Comment:
While I am in favor of avoiding copy/pasting comments for reexported functions, I do think it is more practical from dev point of view to be able to see what a function does immediately, e.g. by hovering (instead of having to click another link).

I would be in favor of having a macro (or some other tool) that takes a path to the reexported function, and copy/pastes the comment for us, wdyt? :thinking:

Reply:
I would say too much work as we probably will have many `#[public]` impls for a smart contract.